app/code/community/Bolt/Boltpay/Helper/Api.php

```diff
@@ -263,8 +263,25 @@ public function createOrder($reference, $sessionQuoteId = null, $isAjaxRequest =
         $service = Mage::getModel('sales/service_quote', $immutableQuote);
 
         try {
+            ///////////////////////////////////////////////////////
+            /// These values are used in the observer after successful
+            /// order creation
+            ///////////////////////////////////////////////////////
+            Mage::getSingleton('core/session')->setBoltTransaction($transaction);
+            Mage::getSingleton('core/session')->setBoltReference($reference);
+            Mage::getSingleton('core/session')->setWasCreatedByMagento($isAjaxRequest);
+            ///////////////////////////////////////////////////////
+
             $service->submitAll();
         } catch (Exception $e) {
+            ///////////////////////////////////////////////////////
+            /// Unset session values set above
+            ///////////////////////////////////////////////////////
+            Mage::getSingleton('core/session')->unsBoltTransaction();
+            Mage::getSingleton('core/session')->unsBoltReference();
+            Mage::getSingleton('core/session')->unsWasCreatedByMagento();
+            ///////////////////////////////////////////////////////
```

Comment:
I recommend moving this three lines to the separate method.

Reply:
My goals are:
First, I recommended to move the logic to separate method where we can manage what we want to unset.
Second, it will useful for PHPUnit test to cover it.

I thought so. Basically, you can skip my recommendation and maybe it would be better to refactor it when it needed.

Reply:
@YevhenShBolt Feel free to refactor it when you think necessary.  When I thought of how the code would look after this, I cannot justify such a change myself.